www/roster/views/actions/memstat.json.rb

```diff
@@ -1,42 +1,121 @@
-  else
-    raise Exception.new("invalid action #{action.inspect}")
-  end
+# update members.txt only for secretary actions
+if updmem
+  ASF::SVN.multiUpdate_ members_txt, message, env, _ do |text|
+    # default command is empty
+    command = ""
+    # remove user's entry
+    unless text.sub! entry, '' # e.g. if the workspace was out of date
+      raise Exception.new("Failed to remove existing entry -- try refreshing")
+    end
+
+    emeritus_url = ASF::SVN.svnurl('emeritus')
+    emeritus_request_url = ASF::SVN.svnurl('emeritus-requests-received')
+    emeritus_reinstated_url = ASF::SVN.svnurl('emeritus-reinstated')
 
-  # perform the insertion
-  text.insert index, entry
+    # determine where to put the entry
+    if @action == 'emeritus'
+      index = text.index(/^\s\*\)\s/, text.index(/^Emeritus/))
+      entry.sub! %r{\s*/\* deceased, .+?\*/},'' # drop the deceased comment if necessary
+      # if pending emeritus request was found, move it to emeritus
+      extra = []
+      # If emeritus request was found, move it to emeritus
+      filename = ASF::EmeritusRequestFiles.extractfilename(@emeritusfileurl)
+      if filename
+        emeritus_destination_url = File.join(emeritus_url, filename)
+        extra << ['mv', @emeritusfileurl, emeritus_destination_url]
+      end
+    elsif @action == 'active'
+      index = text.index(/^\s\*\)\s/, text.index(/^Active/))
+      entry.sub! %r{\s*/\* deceased, .+?\*/},'' # drop the deceased comment if necessary
+      # if emeritus file was found, move it to emeritus-reinstated
+      filename = ASF::EmeritusFiles.extractfilename(@emeritusfileurl)
+      if filename
+        emeritus_destination_url = File.join(emeritus_reinstated_url, filename)
+        extra << ['mv', @emeritusfileurl, emeritus_destination_url]
+      end
+    elsif @action == 'deceased'
+      index = text.index(/^\s\*\)\s/, text.index(/^Deceased/))
+      entry.sub! %r{\n}, " /* deceased, #{@dod} */\n" # add the deceased comment
+    end
 
-  # save the updated text
-  ASF::Member.text = text
+    # perform the insertion
+    text.insert index, entry
 
-  # return the updated (and normalized) text
-  ASF::Member.text
+    # return the updated (and normalized) text and extra svn command
+    [text, extra]
+  end
+end
+
+# Owner operations
+if @action == 'rescind_emeritus'
+  emeritus_rescinded_url = ASF::SVN.svnurl('emeritus-requests-rescinded')
+  pw = " --password #{env.password}" if env.password
+  credentials = "--username #{env.user}#{pw}"
+  command = "svn mv #{credentials} -m \"#{message}\" #{@emeritusfileurl} #{emeritus_rescinded_url}"
+  _.system command
+elsif @action == 'request_emeritus'
+  # Create mail to secretary requesting emeritus
+  FOUNDATION_URL = ASF::SVN.svnurl('foundation')
+  EMERITUS_TEMPLATE_URL = File.join(FOUNDATION_URL, 'emeritus-request.txt').untaint
+  template, err =
+    ASF::SVN.svn('cat', EMERITUS_TEMPLATE_URL, {user: $USER.dup.untaint, password: $PASSWORD.dup.untaint})
```

Comment:
Could use {env: env} instead of user/password

Reply:
done